src/crimson/os/seastore/onode_manager/simple-fltree/onode_node.cc

```diff
@@ -0,0 +1,569 @@
+size_state_t node_t<BlockSize, N, NodeType>::size_state(uint16_t size) const
+{
+  if (size > capacity()) {
+    return size_state_t::overflow;
+  } else if (size < capacity() / 2) {
+    return size_state_t::underflow;
+  } else {
+    return size_state_t::okay;
+  }
+}
+
+template<size_t BlockSize, int N, ntype_t NodeType>
+bool node_t<BlockSize, N, NodeType>::is_underflow(uint16_t size) const
+{
+  switch (size_state(size)) {
+  case size_state_t::underflow:
+    return true;
+  case size_state_t::okay:
+    return false;
+  default:
+    assert(0);
+    return false;
+  }
+}
+
+template<size_t BlockSize, int N, ntype_t NodeType>
+int16_t node_t<BlockSize, N, NodeType>::size_with_key(unsigned slot,
+                                                      const ghobject_t& oid) const
+{
+  if constexpr (item_in_key) {
+    return capacity();
+  } else {
+    // the size of fixed key does not change
+    [[maybe_unused]] const auto& [prefix, suffix] = key_at(slot);
+    return capacity() + partial_key_t::size_from(oid) - suffix.size();
+  }
+}
+
+template<size_t BlockSize, int N, ntype_t NodeType>
+ordering_t node_t<BlockSize, N, NodeType>::compare_with_slot(unsigned slot,
+                                                             const ghobject_t& oid) const
+{
+  const auto& [prefix, suffix] = key_at(slot);
+  if (auto result = prefix.compare(oid); result != ordering_t::equivalent) {
+    return result;
+  } else {
+    return suffix.compare(oid);
+  }
+}
+
+/// return the slot number of the first slot that is greater or equal to
+/// key
+template<size_t BlockSize, int N, ntype_t NodeType>
+std::pair<unsigned, bool> node_t<BlockSize, N, NodeType>::lower_bound(const ghobject_t& oid) const
+{
+  unsigned s = 0, e = count;
+  while (s != e) {
+    unsigned mid = (s + e) / 2;
+    switch (compare_with_slot(mid, oid)) {
+    case ordering_t::less:
+      s = ++mid;
+      break;
+    case ordering_t::greater:
+      e = mid;
+      break;
+    case ordering_t::equivalent:
+      assert(mid == 0 || mid < count);
+      return {mid, true};
+    }
+  }
+  return {s, false};
+}
+
+template<size_t BlockSize, int N, ntype_t NodeType>
+uint16_t node_t<BlockSize, N, NodeType>::size_of_item(const ghobject_t& oid,
+                                                      const item_t& item)
+{
+  if constexpr (item_in_key) {
+    return sizeof(key_t);
+  } else {
+    return (sizeof(key_t) +
+            partial_key_t::size_from(oid) + size_of(item));
+  }
+}
+
+template<size_t BlockSize, int N, ntype_t NodeType>
+bool node_t<BlockSize, N, NodeType>::is_overflow(const ghobject_t& oid,
+                                                 const item_t& item) const
+{
+  return free_space() < size_of_item(oid, item);
+}
+
+template<size_t BlockSize, int N, ntype_t NodeType>
+bool node_t<BlockSize, N, NodeType>::is_overflow(const ghobject_t& oid,
+                                                 const OnodeRef& item) const
```

Comment:
Not sure I get why this overload exists.  I think it's only valid when leaf, so item_t & above would be onode_t  & anyway?

Reply:
yes, this method is only used by the not-yet-posted async part of onode tree:
```c++
  using read_ertr = TransactionManager::read_extent_ertr;
  template<int UpN, class Item>
  read_ertr::future<update_t> insert_at(unsigned slot,
                                        const ghobject_t& oid,
                                        Item item,
                                        InnerNode<BlockSize, UpN>* parent,
                                        unsigned whoami,
                                        Transaction& txn)
  {
    if (me->is_overflow(oid, item)) {
      return split_with(oid, parent, whoami, slot, txn,
        [&](auto& mutable_copy, unsigned n) {
          mutable_copy.mutate(delta_t::insert_item(n, oid, item));
        });
    } else {
      auto mutable_copy = duplicate(txn);
      mutable_copy.mutate(delta_t::insert_item(slot, oid, item));
      return read_ertr::make_ready_future<update_t>();
    }
  }
```

where, if `Item` is `const OnodeRef&`, the node must be a leaf node. otherwise `item_t` should be `child_addr_t`. `onode_t` is used when replaying delta. i could add an `enable_if` so this method is only available for leaf node if you think it's helps.